### ISSUE TITLE
Allowing Scrapy to run without a reactor

### ISSUE THREAD

**Issue by albertedwardson** [scrapy/scrapy#6219]
Do you guys have/had thoughts for migrating to native asyncio? 
What are your thoughts about Twisted? Isn’t it feels deprecated?

**Comment by Gallaecio** [comment]
Twisted works well and is actively developed.

Given that, and the fact that [you can use asyncio code with Scrapy](https://docs.scrapy.org/en/latest/topics/asyncio.html), I am not sure there is a point to migrating away from Twisted.

**Comment by abebus** [comment]
Yes, but it feels somewhat like a crutch, because conversion of futures to defers sometimes can be unpredictable. 
I've had cases where scrapy engine closes all of its resources  before some asyncio tasks where finished.  Soon I will try reproduce this behaviour and provide as a case.

**Comment by wRAR** [comment]
I had a thought about replacing more of the native Deferred code and especially inlineCallbacks code with async def functions, as is recommended by Twisted itself, but rewriting the Scrapy core to remove Twisted completely sounds very hard.

**Comment by dream2333** [comment]
Migrating Scrapy from Twisted to asyncio was not an easy task. It requires a lot of work, including rewriting most of the code, and may also introduce new bugs. in addition, there are some functional differences between Twisted and asyncio, which may result in some features not being implemented on asyncio

**Comment by shraddhakannurkar** [comment]
The perception of Twisted as being deprecated might come from the fact that asyncio is now part of the Python standard library and has gained a lot of traction, leading to a preference for asyncio in new projects. However, Twisted is still actively maintained and has a loyal user base. It's also worth noting that Twisted has been working on integrating with asyncio through the twisted.internet.asyncioreactor module, which allows Twisted to run on top of asyncio's event loop, giving developers the best of both worlds in some cases.

**Comment by dream2333** [comment]
> The perception of Twisted as being deprecated might come from the fact that asyncio is now part of the Python standard library and has gained a lot of traction, leading to a preference for asyncio in new projects. However, Twisted is still actively maintained and has a loyal user base. It's also worth noting that Twisted has been working on integrating with asyncio through the twisted.internet.asyncioreactor module, which allows Twisted to run on top of asyncio's event loop, giving developers the best of both worlds in some cases.

Yes, maybe in the near future we'll be able to migrate more parts to asyncio, as we did from python2 to python3

**Comment by wRAR** [comment]
> Yes, maybe in the near future we'll be able to migrate more parts to asyncio, as we did from python2 to python3

I don't think it's possible to "migrate more parts to asyncio" except for what I've already described, but maybe you have some ideas?

**Comment by wRAR** [comment]
This now looks easier than I thought initially, but more on that later. I think there are different grades of "migrating to native asyncio":

1. Running the asyncio event loop seamlessly and being able to use services provided by it in the user code. The benefits are obvious. This is possible for several years, since 2.0, simply by setting `TWISTED_REACTOR` to the asyncio value. That value is the default for new projects since 2.7 and for all projects since 2.13.
2. Being able to write coroutine-based user code and not even learn about Deferreds and other Twisted APIs. The benefits are obvious. We are not there yet, but you can write many kinds of functions in that way, and more functions if you add trivial conversion code. We improved the related docs in 2.13 and plan to make more APIs usable directly in coroutine-based code. 

[the points above are about writing user code; I believe many of the following points are about the codebase and are irrelevant to many end users except for less tangible things like performance and dependency management]

3. Using the `async def` syntax instead of various Twisted-based syntaxes in internal Scrapy APIs. The benefits are mostly for Scrapy contributors, and they are huge in some places (for `addCallback`-based code; just moving it to `inlineCallbacks` brings almost the same benefits) and noticeable in general. I don't think this is beneficial to users, though there may be slight performance improvements (but also slight performance hits due to added conversion layers). We started doing this in 2.13.
4. Using asyncio-based primitives (`sleep()`, `Future` etc.) instead of Twisted ones directly in the Scrapy code. The benefits are again mostly for Scrapy contributors, and the impact to users is again only about performance and may be negligible except for some edge cases. This is impossible as long as we support non-asyncio values of `TWISTED_REACTOR` (this could change in the future!), unless the code in question is only in the asyncio-enabled code path.
5. Using asyncio-based utilities instead of Twisted-based ones (including 3rd-party libraries, `aiohttp` etc.). This may directly affect users, because different libraries can have different feature sets, performance and quirks. The biggest affected part of Scrapy is its HTTP client (others are the FTP client, the SMTP client and a couple of less interesting parts). It's pluggable and it's possible to write a replacement that uses `aiohttp` right now.
6. Not running the Twisted reactor at all, running the asyncio event loop directly and not relying on any Twisted features that require the reactor. There are many benefits to this, mostly for users that want to run Scrapy as a library, integrating it into something else, and for reducing the amount of parts involved. This requires the previous grade but not necessarily other grades, as one can still use Deferreds without the reactor. There are certain Twisted primitives that can't be used though.
7. Not importing Twisted in the Scrapy code. If the code that requires reactor is removed, the remaining imports would be mostly for the Deferreds. The main benefit here is not having a Twisted dependency.

I think the original question in this ticket and similar user questions we receive from time to time assume either grade 7 ("do not use Twisted at all") or grades 3+4+6 ("use the modern stuff internally") without thinking about 5 (because, frankly, both `HTTP11DownloadHandler` and the fact that it's built upon `twisted.web.client.Agent` are hidden implementation details for most users and I can't say it's a bad thing). But even though 7 makes sense as a final goal, it's not possible to do that while maintaining backward compatibility and if we break it, most users will just keep the old Scrapy version. So the only thing we can do is adding some new and beneficial features while keeping support for the existing code at least with a deprecation period. And I now see how can this be done. While these grades are not a roadmap, they can resemble one. As I mentioned, we already started work for 1-3 and 4 is impossible and so is out of scope. 5 for pluggable things (and most or all are pluggable) can be done gradually. And, assuming a non-Twisted HTTP+HTTPS client, 6 is possible with minor code changes as an optional mode of running Scrapy.

Thus, it should already possible to do this:
- conditionally replace all usages of stuff like `reactor.callLater` in the Scrapy core with asyncio-based ones (I have a PoC for this);
- write an aiohttp-based HTTP client (we have an old PoC for this and a couple of implementations in the wild);
- add some global flag that denotes a "reactorless" mode;
- mark less important optional components as unsupported in this new mode;
- and write new runners/modify existing ones to support this new mode (I have a PoC for a `CrawlerRunner` counterpart).

This, I believe, will allow sharing all 3 modes in the same codebase (using a non-asyncio reactor, using the asyncio reactor, not using a reactor), with a caveat that the 3rd mode is less tested and has a smaller feature set.

**Comment by dream2333** [comment]
> 4\. primitives

听起来很不错，但是如何为实现这个目标做贡献呢？

> This now looks easier than I thought initially, but more on that later. I think there are different grades of "migrating to native asyncio":
> 
> 1. Running the asyncio event loop seamlessly and being able to use services provided by it in the user code. The benefits are obvious. This is possible for several years, since 2.0, simply by setting `TWISTED_REACTOR` to the asyncio value. That value is the default for new projects since 2.7 and for all projects since 2.13.
> 2. Being able to write coroutine-based user code and not even learn about Deferreds and other Twisted APIs. The benefits are obvious. We are not there yet, but you can write many kinds of functions in that way, and more functions if you add trivial conversion code. We improved the related docs in 2.13 and plan to make more APIs usable directly in coroutine-based code.
> 
> [the points above are about writing user code; I believe many of the following points are about the codebase and are irrelevant to many end users except for less tangible things like performance and dependency management]
> 
> 3. Using the `async def` syntax instead of various Twisted-based syntaxes in internal Scrapy APIs. The benefits are mostly for Scrapy contributors, and they are huge in some places (for `addCallback`-based code; just moving it to `inlineCallbacks` brings almost the same benefits) and noticeable in general. I don't think this is beneficial to users, though there may be slight performance improvements (but also slight performance hits due to added conversion layers). We started doing this in 2.13.
> 4. Using asyncio-based primitives (`sleep()`, `Future` etc.) instead of Twisted ones directly in the Scrapy code. The benefits are again mostly for Scrapy contributors, and the impact to users is again only about performance and may be negligible except for some edge cases. This is impossible as long as we support non-asyncio values of `TWISTED_REACTOR` (this could change in the future!), unless the code in question is only in the asyncio-enabled code path.
> 5. Using asyncio-based utilities instead of Twisted-based ones (including 3rd-party libraries, `aiohttp` etc.). This may directly affect users, because different libraries can have different feature sets, performance and quirks. The biggest affected part of Scrapy is its HTTP client (others are the FTP client, the SMTP client and a couple of less interesting parts). It's pluggable and it's possible to write a replacement that uses `aiohttp` right now.
> 6. Not running the Twisted reactor at all, running the asyncio event loop directly and not relying on any Twisted features that require the reactor. There are many benefits to this, mostly for users that want to run Scrapy as a library, integrating it into something else, and for reducing the amount of parts involved. This requires the previous grade but not necessarily other grades, as one can still use Deferreds without the reactor. There are certain Twisted primitives that can't be used though.
> 7. Not importing Twisted in the Scrapy code. If the code that requires reactor is removed, the remaining imports would be mostly for the Deferreds. The main benefit here is not having a Twisted dependency.
> 
> I think the original question in this ticket and similar user questions we receive from time to time assume either grade 7 ("do not use Twisted at all") or grades 3+4+6 ("use the modern stuff internally") without thinking about 5 (because, frankly, both `HTTP11DownloadHandler` and the fact that it's built upon `twisted.web.client.Agent` are hidden implementation details for most users and I can't say it's a bad thing). But even though 7 makes sense as a final goal, it's not possible to do that while maintaining backward compatibility and if we break it, most users will just keep the old Scrapy version. So the only thing we can do is adding some new and beneficial features while keeping support for the existing code at least with a deprecation period. And I now see how can this be done. While these grades are not a roadmap, they can resemble one. As I mentioned, we already started work for 1-3 and 4 is impossible and so is out of scope. 5 for pluggable things (and most or all are pluggable) can be done gradually. And, assuming a non-Twisted HTTP+HTTPS client, 6 is possible with minor code changes as an optional mode of running Scrapy.
> 
> Thus, it should already possible to do this:
> 
> * conditionally replace all usages of stuff like `reactor.callLater` in the Scrapy core with asyncio-based ones (I have a PoC for this);
> * write an aiohttp-based HTTP client (we have an old PoC for this and a couple of implementations in the wild);
> * add some global flag that denotes a "reactorless" mode;
> * mark less important optional components as unsupported in this new mode;
> * and write new runners/modify existing ones to support this new mode (I have a PoC for a `CrawlerRunner` counterpart).
> 
> This, I believe, will allow sharing all 3 modes in the same codebase (using a non-asyncio reactor, using the asyncio reactor, not using a reactor), with a caveat that the 3rd mode is less tested and has a smaller feature set.

I'm particularly interested in the ongoing migration to native asyncio. Are there any areas or issues related to this that would be suitable for a new contributor? Any tips or resources would be greatly appreciated!

**Comment by wRAR** [comment]
@dream2333 while many of the tasks I currently see are either too vague or too complex, I think the following can help:

- testing the just merged `AsyncCrawlerRunner` (#6796)
- adding some basic tests for `ExecutionEngine.download()` that will later help with #6791
- after the aiohttp handler is merged (#6707 / likely #6798), adding more features and tests to it to help reach the feature parity with the default one
- #6792 (the main part of that is refactoring `Crawler` to keep `crawl()` working as before while making `crawl_async()` check the reactor value)
- maybe #6790, the code implementation is likely simple but I'm not sure what should be the interface, as it will inherit `crawl()` from `AsyncCrawlerRunner` (maybe this is easy too, I just haven't thought this through)

**Comment by glyph** [comment]
I can see reasons why you might want to be able to run without Twisted, but I hope that we could help address issues that would make keeping it as a dependency more appealing.

Some of the problems cited here could be addressed by better integration between Twisted and asyncio, and specifically integrating the asyncio loop in the opposite direction; specifically, addressing this issue: https://github.com/twisted/twisted/issues/9968

**Comment by glyph** [comment]
It sounds like maybe we should make another try at automatically converting from Futures (and other awaitables) to Deferreds and vice versa though, as that is a big pain point discussed above?

**Comment by wRAR** [comment]
Hi @glyph!

From my perspective the biggest problems and reasons for a Scrapy user to not want Twisted are:

1. `ReactorNotRestartable`. This alone prevents [several big and popular use cases](https://stackoverflow.com/search?q=%5Bscrapy%5D+reactornotrestartable) which boil down to running a spider several times in succession in the same process without using subprocesses (e.g. notebooks, running a spider inside a Django view, running it with a Python-based scheduler). 
2. Deferred-based APIs. People want to write `async def`, use `await` and not learn what are Deferreds. Hopefully this is no longer needed with very recent Scrapy in simple cases, but can still be needed if you need to interact not just with Scrapy APIs but with Twisted ones too. Also I imagine that people consider `maybe_deferred_to_future()` ugly (I do) and would prefer awaiting on Deferreds directly (I would, when I need to do it).
3. The unfortunate interaction between Windows, Twisted, proactor/selector, loop policies and especially scrapy-playwright (but even without this last component I don't expect Scrapy to work on Python 3.14 on Windows now).

Additionally, what was the actual push for me to start investigating possible Scrapy+Twisted changes was my attempt to integrate Scrapy code and `aiohttp` servers and clients in a single test suite which led me to discover the limitations of `pytest-twisted` with its single reactor and thus single asyncio loop per test session (and also that `pytest-twisted` [may be no longer worked on](https://github.com/pytest-dev/pytest-twisted/issues/188#issuecomment-2679004857)), which I assume is another consequence of `ReactorNotRestartable`.

Still, we are not planning to move on from Twisted as a library as we are not currently planning to deprecate using a Twisted reactor or even non-asyncio Twisted reactors. Also, as we didn't do any benchmarks neither for reactor vs reactorless nor for our default HTTP/1.1 downloader vs alternative downloaders, we don't know what will be the performance differences between these options.

**Comment by glyph** [comment]
> Still, we are not planning to move on from Twisted as a library as we are not currently planning to deprecate using a Twisted reactor or even non-asyncio Twisted reactors. Also, as we didn't do any benchmarks neither for reactor vs reactorless nor for our default HTTP/1.1 downloader vs alternative downloaders, we don't know what will be the performance differences between these options.

Great. I hope we can work together on getting these three issues addressed, thanks for sharing them! I am unfortunately personally quite busy at the moment, but I'll try to get these filed upstream and addressed as much as possible, eventually.

**Comment by albertedwardson** [comment]
may be related: https://github.com/twisted/twisted/pull/1263

**Comment by GeorgeA92** [comment]
@glyph 
> `ReactorNotRestartable.` This alone prevents [several big and popular use cases](https://stackoverflow.com/search?q=%5Bscrapy%5D+reactornotrestartable) which boil down to running a spider several times in succession in the same process without using subprocesses (e.g. notebooks, running a spider inside a Django view, running it with a Python-based scheduler).

Around 1y ago I've achieved some success on overriding `twisted.internet.asyncioreactor.AsyncioSelectorReactor` and `twisted.internet.selectreactor.SelectReactor` to be able to start/stop twisted reactor multiple times in the same process, on env with a bit earlier `twisted` versions https://github.com/scrapy/scrapy/issues/6742
I'd prefer to have something similar supported on twisted level

**Comment by wRAR** [comment]
Released with 2.15.0.
